### ISSUE TITLE
RepoURL in git generator fails to template from cluster generator

### ISSUE THREAD

**Issue by aperullo** [argoproj/argo-cd#15625]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

A git generator's `repoURL` is failing to be templated properly from the result of a cluster generator.

I have a situation where the same repo is exposed under two different urls. There are two clusters, internal and external. Internal can only access the git repo at `https://myrepo.internal/gitops.git` and the other can only access it at  `https://myrepo.external/gitops.git`. In order to use the same repo in each cluster I need to change the repoURL in the git generator, but it doesn't.

**To Reproduce**

Use an ApplicationSet similar to the following:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: application-maker
spec:
  generators:
    - clusters:
        selector:
          matchLabels:
            cluster: internal
          values:
            repoURL: https://myrepo.internal/gitops.git
    - clusters:
        selector:
          matchLabels:
            cluster: external
          values:
            repoURL: https://myrepo.external/gitops.git
    - git:
        repoURL: '{{values.repoURL}}'
        revision: main
        files:
          - path: '*/application.yaml'

  template:
    metadata:
      name: '{{path.basenameNormalized}}'
    spec:
      project: default
      source:
        repoURL: '{{values.repoURL}}'
        targetRevision: HEAD
        path: '{{path}}'
      destination:
        server: 'https://kubernetes.default.svc'
        namespace: default
```

**Expected behavior**

The ApplicationSet should result in one application per path in the git repo, with the url matching whether it was internal or external.

Instead it results in 
```
rpc error: code = Internal desc = unable to resolve git revision : repository not found
```

**Screenshots**

**Version**
`argocd: v2.8.0+804d4b8`

**Comment by crenshaw-dev** [comment]
The `spec.generators` list produces the output of each generator, one by one, and concatenates the output. No generator is aware of any other generators.

I think you're attempting to do something that the `matrix` generator does. I'd try something like this:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: application-maker
spec:
  goTemplate: true
  generators:
    - matrix:
        generators:
          - clusters:
              values:
                repoURL: '{{ (eq .metadata.labels.cluster "internal") | ternary "https://myrepo.internal/gitops.git" "https://myrepo.external/gitops.git" }}'
          - git:
              repoURL: '{{.values.repoURL}}'
              revision: main
              files:
                - path: '*/application.yaml'
  template:
    metadata:
      name: '{{.path.basenameNormalized}}'
    spec:
      project: default
      source:
        repoURL: '{{.values.repoURL}}'
        targetRevision: HEAD
        path: '{{.path.path}}'
      destination:
        server: 'https://kubernetes.default.svc'
        namespace: default
```

**Comment by aperullo** [comment]
Hey thank you for the reply!  I tried a whole bunch of gnarly matrix and merge generator combinations before this and all of them had some problem.

But I tried the one you suggested just in case and it actually crashes the applicationset controller! I updated to the latest argocd to make sure I didn't miss a fix.

version: `argocd: v2.8.4+c279299` 

The environment I have is isolated in such a way that I can't paste the full stack trace but I'll put as I can reasonably type by hand.

```
panic: reflect.Value.SetMapIndex: value of type interface {} is not assignable to type string

/go/pkg/mod/github.com/imdario/mergo@v0.3.16/merge.go:212
/go/pkg/mod/github.com/imdario/mergo@v0.3.16/merge.go:153
/go/pkg/mod/github.com/imdario/mergo@v0.3.16/merge.go:395
...
/go/src/github.com/argoproj/argo-cd/applicationset/controllers/generators/matrix.go:67
```

**Comment by crenshaw-dev** [comment]
Look at me, scaring up bugs.

Do you have the ability to run this and grab the log output? https://github.com/argoproj/argo-cd/pull/15635

**Comment by aperullo** [comment]
Yep I can do that, it'll be a little bit though, I'll ping you when I have something. 

Thanks so much for looking into this.

**Comment by aperullo** [comment]
@crenshaw-dev I have good news. This issue appears to have been fixed sometime between v2.8.4 and v2.9.0 (which I think is what your branch is based on). I was unable to recreate it on your `merge-logging` branch, but was able to on the v2.8.4 image.

Additionally I found it appears to have been related not to the applicationset itself, but to the presence of annotations in the child app (like [here](https://github.com/aperullo/recreate-argocd-15625/blob/main/team-a/application.yaml#L7)

My recreation repo is [here](https://github.com/aperullo/recreate-argocd-15625/tree/main). 

For posterity the full stacktrace was 

```
goroutine 271 [running]:
reflect.Value.assignTo({0x3359f40?, 0xc0004be220?, 0x4fe525?}, {0x3a64238, 0x19}, 0x326c980, 0x0)
	/usr/local/go/src/reflect/value.go:3245 +0x2a5
reflect.Value.SetMapIndex({0x33cf420?, 0xc000f70a20?, 0x0?}, {0x326c980, 0xc0004be1d0, 0x98}, {0x3359f40, 0xc0004be220, 0x94})
	/usr/local/go/src/reflect/value.go:2371 +0x1b6
github.com/imdario/mergo.deepMerge({0x33cf420?, 0xc000f70a20?, 0x0?}, {0x33cd4a0?, 0xc0010b7710?, 0x98?}, 0x3359f40?, 0x2, 0xc0005192a8)
	/go/pkg/mod/github.com/imdario/mergo@v0.3.16/merge.go:212 +0x27a5
github.com/imdario/mergo.deepMerge({0x33cd4a0?, 0xc000f70ba0?, 0xcc00ec2d7f0?}, {0x33cd4a0?, 0xc0010b76e0?, 0x98?}, 0x3359f40?, 0x1, 0xc0005192a8)
	/go/pkg/mod/github.com/imdario/mergo@v0.3.16/merge.go:153 +0x1c74
github.com/imdario/mergo.deepMerge({0x33cd4a0?, 0xc000de48f0?, 0x33cd4a0?}, {0x33cd4a0?, 0xc0010b79b0?, 0x0?}, 0x0?, 0x0, 0xc0005192a8)
	/go/pkg/mod/github.com/imdario/mergo@v0.3.16/merge.go:153 +0x1c74
github.com/imdario/mergo.merge({0x31aebc0, 0xc000de48f0}, {0x33cd4a0, 0xc0010b79b0}, {0x0, 0x0, 0x0?})
	/go/pkg/mod/github.com/imdario/mergo@v0.3.16/merge.go:395 +0x32b
github.com/imdario/mergo.Merge(...)
	/go/pkg/mod/github.com/imdario/mergo@v0.3.16/merge.go:319
github.com/argoproj/argo-cd/v2/applicationset/generators.(*MatrixGenerator).GenerateParams(0xc000382980?, 0xc0004e2410, 0xc000fe0b00)
	/go/src/github.com/argoproj/argo-cd/applicationset/generators/matrix.go:67 +0x419
github.com/argoproj/argo-cd/v2/applicationset/generators.Transform({0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0010086c0, 0x0, 0x0, 0x0}, ...)
	/go/src/github.com/argoproj/argo-cd/applicationset/generators/generator_spec_processor.go:67 +0x7e2
github.com/argoproj/argo-cd/v2/applicationset/controllers.(*ApplicationSetReconciler).generateApplications(_, {{{0x3081398, 0xe}, {0xc000dd00c0, 0x14}}, {{0xc000654768, 0x11}, {0x0, 0x0}, {0xc000cbed80, ...}, ...}, ...})
	/go/src/github.com/argoproj/argo-cd/applicationset/controllers/applicationset_controller.go:502 +0x205
github.com/argoproj/argo-cd/v2/applicationset/controllers.(*ApplicationSetReconciler).Reconcile(0xc000cc4750, {0x4c0f230, 0xc000cab7d0}, {{{0xc000cbed80, 0x6}, {0xc000654768, 0x11}}})
	/go/src/github.com/argoproj/argo-cd/applicationset/controllers/applicationset_controller.go:112 +0x185
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc00051d6b0, {0x4c0f230, 0xc000cab740}, {{{0xc000cbed80?, 0x384be80?}, {0xc000654768?, 0xc000a3d620?}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:114 +0x297
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00051d6b0, {0x4c0f188, 0xc000dd2d70}, {0x3572b80?, 0xc000150020?})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:311 +0x33a
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00051d6b0, {0x4c0f188, 0xc000dd2d70})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:266 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:227 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:223 +0x545
panic: reflect.Value.SetMapIndex: value of type interface {} is not assignable to type string
```

I believe this has been been resolved however and I just need to wait for the next release. Thanks again for the help.
